Mathlib/Analysis/Normed/Group/Ultrametric.lean

```diff
@@ -0,0 +1,88 @@
+import Mathlib.Topology.MetricSpace.Ultra.Basic
+
+/-!
+# Ultrametric normed groups
+
+We give formulations of the ultrametric triangle inequality for ultrametric spaces where the metric
+is induced by a norm; and prove that such groups are nonarchimedean groups (which is *not* true by
+definition in Mathlib – the definition of `NonarchimedeanGroup` does not refer to metrics or norms).
+-/
+
+namespace IsUltrametricDist
+
+/-- Triangle inequality for ulrametric (semi) norms on groups. -/
+@[to_additive "Triangle inequality for ulrametric (semi) norms on additive groups."]
+lemma norm_mul_le_max
+    {G : Type*} [SeminormedGroup G] [IsUltrametricDist G]
+    (x y : G) : ‖x * y‖ ≤ max ‖x‖ ‖y‖ := by
+  simpa only [dist_eq_norm_div, div_inv_eq_mul, div_one, one_mul] using dist_triangle_max x 1 y⁻¹
+
+/--
+In an commutative group with an ultrametric norm, the open balls around 1 are open subgroups.
+-/
+@[to_additive "In an commutative additive group with an ultrametric norm, the open balls around 0
+are open subgroups."]
+def ball_openSubgroup (G : Type*) [SeminormedGroup G] [IsUltrametricDist G]
+    {r : ℝ} (hr : 0 < r) : OpenSubgroup G where
+  carrier := Metric.ball (1 : G) r
+  mul_mem' {x} {y} hx hy := by
+    simp only [Metric.mem_ball, dist_eq_norm_div, div_one] at hx hy ⊢
+    exact (norm_mul_le_max x y).trans_lt (max_lt hx hy)
+  one_mem' := Metric.mem_ball_self hr
+  inv_mem' := by simp only [Metric.mem_ball, dist_one_right, norm_inv', imp_self, implies_true]
+  isOpen' := Metric.isOpen_ball
+
+/-- A commutative normed group with an ultrametric norm is nonarchimedean. -/
+@[to_additive "A commutative normed additive group with an ultrametric norm is nonarchimedean"]
+instance nonarchimedeanGroup_of_ultrametric {G : Type*} [SeminormedCommGroup G]
+    [IsUltrametricDist G] : NonarchimedeanGroup G := by
+  apply NonarchimedeanGroup.mk
+  simpa only [Metric.mem_nhds_iff] using fun U ⟨ε, hεp, hεU⟩ ↦ ⟨ball_openSubgroup G hεp, hεU⟩
+
+section finsums
+/-!
+## Bounds for the norms of finite sums / products
+-/
+
+variable {G ι : Type*} [SeminormedCommGroup G] [IsUltrametricDist G] [DecidableEq ι]
+/--
+Generalised ultrametric triangle inequality for nonempty finite products in groups with an
+ultrametric norm.
+-/
+@[to_additive "Generalised ultrametric triangle inequality for nonempty finite sums in additive
+groups with an ultrametric norm ."]
+lemma norm_prod_le_of_forall_le_of_nonempty {s : Finset ι} (hs : s.Nonempty) {f : ι → G} {C : ℝ}
+    (hC : ∀ i ∈ s, ‖f i‖ ≤ C) : ‖∏ i ∈ s, f i‖ ≤ C := by
+  induction' s using Finset.induction_on with i t hi ht
+  · exact (hs.ne_empty rfl).elim
+  · rcases t.eq_empty_or_nonempty with rfl | ht'
+    · simp only [insert_emptyc_eq, Finset.prod_singleton]
+      exact hC i (Finset.mem_insert_self i ∅)
+    · rw [Finset.prod_insert hi]
+      refine (norm_mul_le_max _ _).trans (max_le (hC _ (t.mem_insert_self i)) ?_)
+      exact ht ht' (fun j hj ↦ hC _ <| Finset.mem_insert_of_mem hj)
```

Comment:
Do you think it's worthwhile to combine these with my lemmas in #14768 
https://github.com/leanprover-community/mathlib4/pull/14768/files#diff-59aef0da4a062256a000d746f59803a758717edfd3fd33f52b9ed92edafe5d03R169-R182
I know I proved them there for `SeminormedAddGroup` but should be easy to switch them to `SeminormedGroup` and additivize

Reply:
I've multiplicativized #14768 so that this and related lemmas are there